src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/AbstractFlinkTimerInternals.java

```diff
@@ -34,110 +34,93 @@
-
 	@Override
-	public void deleteTimer(com.google.cloud.dataflow.sdk.util.TimerInternals.TimerData timerKey) {
-		K key = element.isTimer() ? (K) element.key() : element.element().getValue().getKey();
-		unregisterTimer(key, timerKey);
+	public Instant currentProcessingTime() {
+		return Instant.now();
 	}
 
-	protected abstract void unregisterTimer(K key, TimerData timerKey);
+	@Override
+	public Instant currentInputWatermarkTime() {
+		return currentInputWatermark;
+	}
 
+	@Nullable
 	@Override
-	public Instant currentProcessingTime() {
-		return Instant.now();
+	public Instant currentSynchronizedProcessingTime() {
+		// TODO
+		return null;
 	}
 
 	@Override
-	public Instant currentWatermarkTime() {
-		return this.currentWatermark;
+	public Instant currentOutputWatermarkTime() {
+		return currentOutputWatermark;
 	}
 
-	private void checkIfValidWatermark(Instant newWatermark) {
-		if (currentWatermark.isAfter(newWatermark)) {
+	private void checkIfValidInputWatermark(Instant newWatermark) {
+		if (currentInputWatermark.isAfter(newWatermark)) {
 			throw new IllegalArgumentException(String.format(
-					"Cannot set current watermark to %s. Newer watermarks " +
+					"Cannot set current input watermark to %s. Newer watermarks " +
 							"must be no earlier than the current one (%s).",
-					newWatermark, this.currentWatermark));
+					newWatermark, currentInputWatermark));
+		}
+	}
+
+	private void checkIfValidOutputWatermark(Instant newWatermark) {
+		if (currentOutputWatermark.isAfter(newWatermark)) {
+			throw new IllegalArgumentException(String.format(
+				"Cannot set current output watermark to %s. Newer watermarks " +
+					"must be no earlier than the current one (%s).",
+				newWatermark, currentOutputWatermark));
 		}
 	}
 
 	public void encodeTimerInternals(DoFn.ProcessContext context,
-									  StateCheckpointWriter writer,
-									  KvCoder<K, VIN> kvCoder,
-									  Coder<? extends BoundedWindow> windowCoder) throws IOException {
+	                                 StateCheckpointWriter writer,
+	                                 KvCoder<K, VIN> kvCoder,
+	                                 Coder<? extends BoundedWindow> windowCoder) throws IOException {
 		if (context == null) {
 			throw new RuntimeException("The Context has not been initialized.");
 		}
 
-		if (element != null && !element.isTimer()) {
-			// create the element coder
-			WindowedValue.WindowedValueCoder<KV<K, VIN>> elementCoder = WindowedValue
-					.getFullCoder(kvCoder, windowCoder);
-
-			CoderTypeSerializer<WindowedValue<KV<K, VIN>>> serializer =
-					new CoderTypeSerializer<>(elementCoder);
-
-			writer.writeByte((byte) 1);
-			writer.serializeObject(element.element(), serializer);
-		} else {
-			// just setting a flag to 0, meaning that there is no value.
-			writer.writeByte((byte) 0);
-		}
-		writer.setTimestamp(currentWatermark);
+		writer.setTimestamp(currentInputWatermark);
+		writer.setTimestamp(currentOutputWatermark);
```

Comment:
That makes sense. No need to use Beam's `StateCheckpointWriter` here for the elements.